Add hidpi patches for gtk+-3.0

lazka opened the pull request:
One in gtk+ to use the system provided scaling factor and one
in cairo to disable CGContexts scaling so we can do our own scaling.

This fixes pixelated rendering in all cases where the gtk pixel cache
is used (treeviews..) and where icons are converted to cairo surfaces.

Any remaining pixelation (DnD, menu icons) was fixed in gtk+3.18

This also adds the gtk patch to gtk+-3.14 as the cairo patch alone breaks
drawing on hidpi setups without the gtk+ counterpart.

----------------

I've shipped these patches for a year now with gtk+3.16 and my hidpi fixes which are also included in gtk+3.18: https://git.gnome.org/browse/gtk+/log/?qt=grep&q=quartz.*hidpi

I haven't found the time to look into upstreaming these.. but I guess it would be good to have working hidpi with gtk-osx.

lazka commented:
To clarify: These patches here are not in gtk+/cairo master. I've only pushed fixes for problems exposed by these two patches in gtk+ master.

HiDPI can be enabled on non-retina displays using the "Graphic Tools for Xcode" afair. It exposes an additional hidpi resolution option in the normal display resolution dialog.

lazka commented:
For how the pixelation looks, here is an example: https://cloud.githubusercontent.com/assets/4942960/8472311/706ab996-20a0-11e5-935e-c2cc4bbeaf1d.png

jralls commented:
>To clarify: These patches here are not in gtk+/cairo master. I've only pushed fixes for problems exposed by these two patches in gtk+ master.

OK, where did they come from? If they're on bugs then include the bug in the patch name so that we know when to take them out.

Also you need to rebase your branch, the PR currently undoes the pango update.

lazka commented:
> OK, where did they come from? If they're on bugs then include the bug in the patch name so that we know when to take them out.

Will do.

> Also you need to rebase your branch, the PR currently undoes the pango update.

oops :/

lazka commented:
>> OK, where did they come from? If they're on bugs then include the bug in the patch name so that we know when to take them out.
>
>Will do.

Added links to the source bugs as xml comment in the module and in the commit message.

>> Also you need to rebase your branch, the PR currently undoes the pango update.
>
>oops :/

fixed now

jralls commented:
Thanks.

You need to cherry-pick the commits from gtk-3-18 to master ASAP so they make it into 3.20!

lazka commented:
Thanks

> You need to cherry-pick the commits from gtk-3-18 to master ASAP so they make it into 3.20!

Sorry, I don't follow. Which commits?

jralls commented:
The two commits that you made in gtk-3.18, that you referenced in your opening message and then said aren't in master.

lazka commented:
I meant I've pushed them during the 3.18 cycle, so they are also in master. Sorry for the confusion :/

jralls commented:
https://bugzilla.gnome.org/show_bug.cgi?id=763269 reports that this breaks gtk-2-24. That shouldn't be a surprise as your comment on the commit points out. Unfortunately gtk2 doesn't implement gdk_quartz_window_ref_cairo_surface so we can't backport the gtk3 patch.

I think we can work around that by moving the cairo patch to gdk_quartz_create_cairo_surface in gdk/quartz/gdkwindow-quartz.c after checking that we got a context (line 294 in master). I don't have a retina monitor so I can't test it. Can you?
